{
 "metadata": {
  "name": "",
  "signature": "sha256:77660e0db4b5d01fa2a6126cd8e487b7f16d5b81df44b8049d9e9f1c9eaed1d3"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Description:\n",
      "\n",
      "* intial dataset setup for deltaGC\n",
      " * dataset consisting of all closed genomes downloaded from NCBIa\n",
      "   * genomes downloaded on 27-Mar-2014\n",
      " * rnammer used to identify all 16S genes\n",
      " \n",
      "* initial investigation of GC content\n",
      " * trying to use GC content of the nucleotide fragment amplified from a gradient fraction sample to see if the fragment was labeled \n",
      " * determining how read GC relates to fragment GC  "
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Downloading genomes\n",
      "\n",
      "Date: 3/27/2014\n",
      "\n",
      "* using 'prokaryote.txt' file from the NCBI genome ftp site\n",
      "\n",
      "    wget ftp://ftp.ncbi.nih.gov/genomes/GENOME_REPORTS/prokaryotes.txt\n",
      "\n",
      "\n",
      "__Table filtering pipeline:__\n",
      "* just closed genomes (1 scaffold per chromosome)\n",
      " * 'Complete' in 'Status' field\n",
      "* need to pull the accession numbers for the most recent version of the genome\n",
      " * accession field: 'Chromosomes/INSDC'\n",
      "* recent: 'Release Date'\n",
      " * need to add taxonomy to table\n",
      "\n",
      "\n",
      "    wget ftp://ftp.ncbi.nih.gov/genomes/GENOME_REPORTS/prok_representative_genomes.txt\n",
      "    \n",
      "    wget ftp://ftp.ncbi.nih.gov/genomes/GENOME_REPORTS/prok_reference_genomes.txt\n",
      "    \n",
      "    \n",
      "## Filtering prokaryotes file\n",
      "\n",
      "* just full chromosomes\n",
      "* also adding taxonomy\n",
      "\n",
      "@ system76-server:/var/seq_data/ncbi_db/genome\n",
      "    \n",
      "    NCBIprokaryoteTableFilter.pl prokaryotes.txt > prokaryotes_filt.txt\n",
      "        * 23504 entries in prokaryotes\n",
      "        * 2864 entries remaining post-filtering\n",
      "         * 36 genomes do not have accession numbers\n",
      "         * 2828 genomes remaining with chromosome accessions\n",
      "\n",
      "__random selection of 1 from each species__\n",
      "\n",
      "* species column in prokaryotes table: 30\n",
      "    \n",
      "@ system76-server:/var/seq_data/ncbi_db/genome\n",
      "    \n",
      "    tblRandomByField.pl -c 30 -header <(perl -ne '@l=split /\\t/; print unless $l[9] eq \"-\"' prokaryotes_filt.txt) > prokaryotes_filt_rand.txt\n",
      "        * filtered out all genomes w/out accession numbers\n",
      "        * 1283 genomes remaining\n",
      "\n",
      "__parsing by domain__\n",
      "\n",
      "    perl -ne '@l=split /\\t/; print if $l[23] =~/bacteria|superkingdom/i && $l[29] !~ /NA/' prokaryotes_filt_rand.txt > prok-bac_filt_rand.txt\n",
      "        * 1162 bacterial genomes\n",
      "    perl -ne '@l=split /\\t/; print if $l[23] =~/archaea|superkingdom/i && $l[29] !~ /NA/' prokaryotes_filt_rand.txt > prok-arc_filt_rand.txt\n",
      "        * 119 archaeal genomes\n",
      "        \n",
      "__downloading all genomes__\n",
      "\n",
      "    screen -S bac -L NCBIprokaryoteTableFilter.pl prok-bac_filt_rand.txt -t 20 -w -d prok-bac-genomes\n",
      "    \n",
      "    screen -S arc -L NCBIprokaryoteTableFilter.pl prok-arc_filt_rand.txt -t 20 -w -d prok-arc-genomes\n",
      "    "
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Running RNAmmer (v1.2) on all genomes\n",
      "\n",
      "Identifying ssu, lsu, and tsu genes in each genome\n",
      "\n",
      "## rnammer run\n",
      "\n",
      "__archaea_ssu__\n",
      "\n",
      "@ system76-server:/var/seq_data/ncbi_db/genome/rnammer/archaea_ssu\n",
      "    \n",
      "    find ../../prok-arc-genomes/ -name \"*fasta\" | perl -pe 's/.+\\/|\\.fasta//g' | xargs -n 1 -I % -P 30 bash -c 'rnammer -S arc -m ssu,lsu,tsu -gff %_rrn.gff -f %_rrn.fna -xml %_rrn.xml < ../../prok-arc-genomes/%.fasta'\n",
      "\n",
      "__bacteria_ssu__\n",
      "\n",
      "@ system76-server:/var/seq_data/ncbi_db/genome/rnammer/bacteria_ssu\n",
      "    \n",
      "    find ../../prok-bac-genomes/ -name \"*fasta\" | perl -pe 's/.+\\/|\\.fasta//g' | xargs -n 1 -I % -P 30 bash -c 'rnammer -S bac -m ssu,lsu,tsu -gff %_rrn.gff -f %_rrn.fna -xml %_rrn.xml < ../../prok-bac-genomes/%.fasta'\n",
      "    \n",
      "\n",
      "## Summarizing the results\n",
      "\n",
      "@ system76-server:/var/seq_data/ncbi_db/genome/rnammer/\n",
      "    \n",
      "    egrep -v \"^#\" archaea_ssu/*gff | grep \"16s_rRNA\" | perl -pe 's/:/\\t/' > summary/archaea_ssu_gff.txt\n",
      "    egrep -v \"^#\" bacteria_ssu/*gff | grep \"16s_rRNA\" | perl -pe 's/:/\\t/' > summary/bacteria_ssu_gff.txt\n",
      "    \n",
      "    egrep -c \"^[^#]\" archaea_ssu/*gff | perl -pe 's/:/\\t/' > summary/archaea_ssu_gff_cnt.txt\n",
      "    egrep -c \"^[^#].+16s_rRNA\" bacteria_ssu/*gff | perl -pe 's/:/\\t/' > summary/bacteria_ssu_gff_cnt.txt\n",
      "\n",
      "\n",
      "    $ cut -f 2 summary/archaea_ssu_gff_cnt.txt | stats_descriptive.pl\n",
      "    1\tmin\t1.00\n",
      "    1\tQ1\t1.00\n",
      "    1\tmean\t1.64\n",
      "    1\tmedian\t1.00\n",
      "    1\tQ3\t2.00\n",
      "    1\tmax\t4.00\n",
      "    1\tstdev\t0.86\n",
      "\n",
      "    cut -f 2 summary/bacteria_ssu_gff_cnt.txt | stats_descriptive.pl\n",
      "    1\tmin\t0.00\n",
      "    1\tQ1\t2.00\n",
      "    1\tmean\t3.68\n",
      "    1\tmedian\t3.00\n",
      "    1\tQ3\t5.00\n",
      "    1\tmax\t15.00\n",
      "    1\tstdev\t2.55\n",
      "\n",
      "    * 4 bacterial lacking identified 16S genes!\n",
      "    bacteria_ssu/Bacteroides_xylanisolvens_XB1A_ssu.gff     0\n",
      "    bacteria_ssu/Candidatus_Phytoplasma_solani_284_09_ssu.gff       0\n",
      "    bacteria_ssu/Faecalibacterium_prausnitzii_L2-6_ssu.gff  0\n",
      "    bacteria_ssu/Streptomyces_rapamycinicus_NRRL_5491_ssu.gff       0\n",
      "        * genomes lengths make sense\n",
      "        * just seem to be lacking the gene\n",
      "        \n"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Determing how many rrn operons are within 5 kb from each other\n",
      "\n",
      "### Need to re-run rnammer and identify ssu & lsu genes\n",
      "\n",
      "__Archaea__\n",
      "\n",
      "@ system76-server:/var/seq_data/ncbi_db/genome/rnammer/archaea_rrn\n",
      "\n",
      "    find ../../prok-arc-genomes/ -name \"*fasta\" | perl -pe 's/.+\\/|\\.fasta//g' | xargs -n 1 -I % -P 30 bash -c 'rnammer -S arc -m ssu,tsu,lsu -gff %_rrn.gff -f %_rrn.fna -xml %_rrn.xml < ../../prok-arc-genomes/%.fasta'\n",
      "    \n",
      "    rnammer_tandem_rrn.pl <(find archaea_rrn/ -name \"*gff\") | tail -n +2 | perl -ne '@l=split /\\t/; print if $l[2] > 0' | less\n",
      "        * 2 genomes (1.7%); both methanogens\n",
      "       \n",
      "\n",
      "__Bacteria__\n",
      "\n",
      "@ system76-server:/var/seq_data/ncbi_db/genome/rnammer/bacteria_rrn\n",
      "\n",
      "    find ../../prok-bac-genomes/ -name \"*fasta\" | perl -pe 's/.+\\/|\\.fasta//g' | xargs -n 1 -I % -P 30 bash -c 'rnammer -S bac -m ssu,tsu,lsu -gff %_rrn.gff -f %_rrn.fna -xml %_rrn.xml < ../../prok-bac-genomes/%.fasta'\n",
      "        * 126 genomes (11.9%)\n",
      "        * mean of 25% tandem (max of 83% tandem)\n"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Dataset summary\n",
      "\n",
      "__Number of genomes__\n",
      "\n",
      "* Number of bacterial genomes (1 rep for each species): 1163\n",
      " * 4 do not have an identified 16S gene\n",
      "* Number of archaeal genomes (1 rep for each species): 119\n",
      "\n",
      "__Number of phyla__\n",
      "\n",
      "@ system76-server:/var/seq_data/ncbi_db/genome\n",
      "\n",
      "    cut -f 25 prok-bac_filt_rand.txt | tail -n +2 | sort -u | less -N\n",
      "        * 29 bacterial phyla\n",
      "    \n",
      "    cut -f 25 prok-arc_filt_rand.txt | tail -n +2 | sort -u | less -N\n",
      "        * 5 archaeal phyla\n",
      "        \n",
      "\n"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Calculating GC content for each genome\n",
      "\n",
      "__bacteria__\n",
      "\n",
      "@ system76-server:~/notebook/deltaGC_notes/data/genome_data/prok-bac-genomes_GC\n",
      "\n",
      "    find ../prok-bac-genomes/ -name \"*fasta\" | xargs -P 5 -I % calcGC.pl -w 6 %  > prok-bac-genomes_GC.txt\n",
      "    \n",
      "__archaea__    \n",
      "    \n",
      "@ system76-server:~/notebook/deltaGC_notes/data/genome_data/prok-arc-genomes_GC\n",
      "\n",
      "    find ../prok-arc-genomes/ -name \"*fasta\" | xargs -P 5 -I % calcGC.pl -w 6 %  > prok-arc-genomes_GC.txt\n",
      "\tmin\t0.13\n",
      "\tQ1\t0.37\n",
      "\tmean\t0.48\n",
      "\tmedian\t0.46\n",
      "\tQ3\t0.61\n",
      "\tmax\t0.74\n",
      "\tstdev\t0.13 \n",
      " \n",
      "    cut -f 2 prok-arc-genomes_Ngaps_GC.txt | stats_descriptive.pl\n",
      "\tmin\t0.27\n",
      "\tQ1\t0.37\n",
      "\tmean\t0.47\n",
      "\tmedian\t0.47\n",
      "\tQ3\t0.56\n",
      "\tmax\t0.67\n",
      "\tstdev\t0.11"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "\n"
     ]
    }
   ],
   "metadata": {}
  }
 ]
}